src/ray/core_worker/core_worker.cc

```diff
@@ -271,16 +274,19 @@ CoreWorker::CoreWorker(const CoreWorkerOptions &options, const WorkerID &worker_
             new rpc::CoreWorkerClient(addr, *client_call_manager_));
       });
 
-  // Register a callback to monitor removed nodes.
+  // Register a callback to monitor add/removed nodes.
   // Note we capture a shared ownership of reference_counter_
   // here to avoid destruction order fiasco between gcs_client and reference_counter_.
-  auto on_node_change = [reference_counter = this->reference_counter_](
+  auto on_node_change = [this, reference_counter = this->reference_counter_](
```

Comment:
c++ noob question: is there any danger with `this` pointer copy? what happens if the lambda is called after the instance is destructed? Is it safer to do `this->num_alive_nodes`?

Reply:
we don't destruct the core worker until io service is stopped / drained, so it should be safe

Reply:
Ok. hmm nit: use safer mechanism to defend against bugs intro'd by future refactors